File: kotlin-node/karakum/src/jsMain/resources/plugins/convertKeyUnions.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        ts.isTypeAliasDeclaration(node)
        && (
            node.name.text === "KeyType"
            || node.name.text === "KeyFormat"
        )
        && ts.isUnionTypeNode(node.type)
        && sourceFileName.endsWith("crypto.d.ts")
    ) {
        const name = render(node.name)

        const entries = node.type.types
            .filter(type => ts.isLiteralTypeNode(type))
            .map(type => type.literal)
            .filter(literal => ts.isStringLiteral(literal))
            .map(literal => {
                const value = literal.text
                const key = karakum.identifier(value)
                return [key, value]
            })

        const keys = entries.map(([key]) => key)

        const body = keys
            .map(key => `sealed interface ${key} : ${name}`)
            .join("\n")

        const companionBody = entries
            .map(([key, value]) => (
                `
@seskar.js.JsValue("${value}")
val ${key}: ${key}
                    `.trim()
            ))
            .join("\n")

        return `
sealed external interface ${name} {
${body}

companion object {
${companionBody}
}
}
        `
    }


    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertKeyUnions.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.util.camelize
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.*

private val keyGroups = setOf(
    "mlDsa",
    "mlKem",
    "slhDsa",
)

val convertKeyUnions = createPlugin { node, _, render ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("crypto.d.ts"))

        ensure(isTypeAliasDeclaration(node))
        ensure(
            node.name.text == "KeyType"
                    || node.name.text == "KeyFormat"
        )

        val union = node.type
        ensure(isUnionTypeNode(union))

        val name = render(node.name)

        val entries = union.types.asArray()
            .mapNotNull {
                nullable {
                    ensure(isLiteralTypeNode(it))

                    val literal = it.literal
                    ensure(isStringLiteral(literal))

                    val value = literal.text
                    val key = camelize(value)

                    key to value
                }
            }

        val keys = entries.map { (key) -> key }

        val typeGroups = keyGroups.joinToString("\n") {
            "sealed interface $it : $name"
        }

        val types = keys.joinToString("\n") { key ->
            val group = keyGroups.find { key.startsWith(it) }

            "sealed interface $key : $name${ifPresent(group) { ", $group" }}"
        }

        val companionBody = entries.joinToString("\n") { (key, value) ->
            """
                @seskar.js.JsValue("$value")
                val $key: $key
            """.trimIndent()
        }

        """
            sealed external interface $name {
            $typeGroups
            $types

            companion object {
            $companionBody
            }
            }
        """.trimIndent()
    }
}
